### PR TITLE
Pile overlay revamp

### PR DESCRIPTION
**Pull request by mousewax**
Against my better judgement I am still doing stuff. This is to partially address a very old Issue.
https://github.com/ArnoldSmith86/virtualtabletop/issues/53

It's not fully polished but it works with no errors from the back-en.

The split section has three inputs. You can change either side of the split or use the slider to change the split. 

If a user inputs values greater than the number of cards it sets the field to the number of cards. Less than 0 returns 0.

image

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:4019/pr-test (or any other room on that server)

**Comment by 96LawDawg** [comment]
Good start.  Everything works like it should.  As you say, just needs some polish.  I don't suppose it is possible to have the numbers on the split pile change while you drag the slider?  Right now you can increment the numbers using scroll bars.  Or you can drag the slider.  But the slider doesn't change the numbers until you let go.

**Comment by mousewax** [comment]
The numbers are updated from a `change` event on the slider. The event does not fire until the user lets go. It's the same deal when you hold down the up or down icons on the number inputs.

Though some UI libraries seem to have ways to get around that. I will do some research.

**Comment by mousewax** [comment]
Ta-dah!

**Comment by 96LawDawg** [comment]
Fantastic!  And fast.  But can you do the same thing with the scrollbar?  In other words "The little lights are not twinkling, Clark."  (Movie playing in the background in the other room now.  Hopefully you get the reference).

**Comment by bjalder26** [comment]
I think TestCafe will need to be updated to account for the modal.

**Comment by 96LawDawg** [comment]
Two more thoughts.  Should the slider and scrollbar start at 1 and go to n-1?  There is no point in splitting a pile of 0 or a pile of all the cards.  Sure, you can do that on production now, but should we allow that?

Also, you can enter a decimal number.  It rounds up.  That happens on production as well.  Maybe prevent decimals?

**Comment by 96LawDawg** [comment]
Perhaps an unintended side effect ... You can't backspace to clear out the current entry and just type something new in.  It starts at 26.  Backspace and it clears the 2.  Backspace again and it reduces to 1.  Then you have to type 15 then delete just the 1 to get 5.

Also, you can still type 52 and get a 52 card split.

Also, I'm just really nitpicking now trying to find out it can be broken.  I have never ONCE ever used split in an actual game.  So none of this really matters to me.

**Comment by 96LawDawg** [comment]
Another thought.  Could the overlay just be a pop up like pictured in #53 or in the Custom Pile Memo demo at https://virtualtabletop.io/DemoLibrary?  It would be nice if the overlay was very close to where the pile handle was to start with.

**Comment by mousewax** [comment]
> Another thought. Could the overlay just be a pop up like pictured in #53 or in the Custom Pile Menu demo at https://virtualtabletop.io/DemoLibrary? It would be nice if the overlay was very close to where the pile handle was to start with.

*I* cannot do that, but I am positive it can be done by someone with more knowledge than me. 

> You can't backspace to clear out the current entry and just type something new in. It starts at 26. Backspace and it clears the 6. Backspace again and it reduces to 1. Then you have to type 15 then delete just the 1 to get 5.

I cannot change that. Double click/tap the number to highlight it then type a number? It's a mild annoyance but that kind of thing is a normal computer interaction.

File: client/js/widgets/pile.js [hunk 1]
```diff
@@ -75,10 +75,29 @@ class Pile extends Widget {
...
+//       $('#pileOverlay')
+//
+//       const modalTitle = document.createElement('div');
+//       modalTitle.classList.add = 'inputtitle';
+//       var thisField = { type: 'title',text: ${this.handle.textContent} cards+' cards' };
+//       formField(thisField, modalTitle, 'pileTitle')
+//       modalDiv.appendChild(modalTitle);
+
+
+
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:91]: What's the plan here? You marked it as "ready to review" but this doesn't look finished yet. Not that that's a problem; I just don't know how to review this. :smile:

File: client/js/widgets/pile.js [hunk 2]
```diff
@@ -90,34 +109,105 @@ class Pile extends Widget {
...
         batchEnd();
       });
-      $('#pileOverlay').appendChild(flipButton);
+      buttonBar1.appendChild(flipButton);
+
+      const or1label = document.createElement('div');
+      or1label.className = 'inputtext optiontext';
+      const thisOr1Field = { type: 'text',text:'-OR-' };
+      formField(thisOr1Field, or1label, 'pileTitle')
+      $('#pileOverlay > .modal').appendChild(or1label);
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:118]: I don't see an `-OR-` on the overlay. Is this a WIP?

File: client/js/widgets/pile.js [hunk 3]
```diff
@@ -90,34 +109,105 @@ class Pile extends Widget {
...
+      buttonBar2.appendChild(shuffleButton);
+
+
+
+
+      const splitLabel = document.createElement('div');
+      splitLabel.className = 'inputtext optiontext';
+      const thisField = { type: 'text',text:'-OR-' };
+      formField(thisField, splitLabel, 'pileTitle')
+      $('#pileOverlay > .modal').appendChild(splitLabel);
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:143]: Same question for the `-OR-`.

File: client/js/widgets/pile.js [hunk 4]
```diff
@@ -90,34 +109,105 @@ class Pile extends Widget {
...
-      splitInput.min = 0;
-      splitInput.max = childCount;
-      countDiv.prepend(splitInput);
-      const splitLabel = document.createElement('label');
-      splitLabel.textContent = 'Split: ';
-      countDiv.prepend(splitLabel);
+      splitInput.min = 1;
+      splitInput.max = childCount - 1;
+      splitInput.setAttribute("aria-valuemin", 0);
+      splitInput.setAttribute("aria-valuemax", childCount);
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:157]: Are the `aria` values supposed to be different?

File: client/js/widgets/pile.js [hunk 5]
```diff
@@ -90,34 +109,105 @@ class Pile extends Widget {
...
+      });
+      countDiv.appendChild(splitInput);
+
+      const splitInputSlider = document.createElement('input');
+      splitInputSlider.type = 'range';
+      splitInputSlider.value = Math.floor(childCount/2);
+      splitInputSlider.min = 1;
+      splitInputSlider.max = childCount - 1;
+      splitInputSlider.setAttribute("aria-valuemin", 0);
+      splitInputSlider.setAttribute("aria-valuemax", childCount);
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:177]: Same question about `aria`.

File: client/js/widgets/pile.js [hunk 6]
```diff
@@ -90,34 +109,105 @@ class Pile extends Widget {
...
+      });
+      countDiv.appendChild(splitInputSlider);
+
+      const denominatorInput = document.createElement('input');
+      denominatorInput.value = childCount - splitInput.value;
+      denominatorInput.type = 'number';
+      denominatorInput.min = 1;
+      denominatorInput.max = childCount - 1;
+      denominatorInput.setAttribute("aria-valuemin", 0);
+      denominatorInput.setAttribute("aria-valuemax", childCount);
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:191]: Same question. Does `aria` support even make sense? I'm pretty sure that games will never be playable using a screen reader. But I don't know if it's useful for any other circumstances.

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
I didn't look at the CSS. The code looks fine for the most part (see comments).

**Comment by RaphaelAlvez** [comment]
this looks so much better @mousewax do you mind just taking a look at what was commented?

also, any reason why the overlay is offset?

**Review by ArnoldSmith86 — APPROVED** [review]
I addressed my own comments.

Somebody else should probably review this as well. :smile:

**Comment by ArnoldSmith86** [comment]
image

This is what it looks like for me on Firefox/Linux now. Please post screenshots if it looks different for you on any platform.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.  Looks the same to me on all browsers, including Safari mobile.  The only thing I noticed is that if the card count gets really high, like in CAH, then the slider isn't centered when the window opens.  It works fine in Flight with 200 cards, but not in CAH with 500.

CAH

**Pull request merged**